src/clinical/id-generator.ts

```diff
@@ -0,0 +1,263 @@
+ * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT
+ * SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
+ * TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
+ * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER
+ * IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ * ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import axios from 'axios';
+import memoize from 'memoizee';
+import ms from 'ms';
+import { Donor } from './clinical-entities';
+import { ClinicalEntitySchemaNames } from '../common-model/entities';
+import { config } from '../config';
+import { Errors } from '../utils';
+
+interface IdGenerationRequest {
+	programId: string;
+	submitterSpecimenId: string;
+	submitterSampleId: string;
+	submitterPrimaryDiagnosisId: string;
+	submitterFollowUpId: string;
+	submitterTreatmentId: string;
+	submitterDonorId: string;
+	testInterval: string;
+	family_relative_id: string;
+	comorbidityTypeCode: string;
+	entityType: string;
+}
+
+const request = {
+	programId: '-',
+	submitterSpecimenId: '-',
+	submitterSampleId: '-',
+	submitterPrimaryDiagnosisId: '-',
+	submitterFollowUpId: '-',
+	submitterTreatmentId: '-',
+	submitterDonorId: '-',
+	testInterval: '-',
+	family_relative_id: '-',
+	comorbidityTypeCode: '-',
+	entityType: '-',
+};
+
+export interface PartialDonor extends Partial<Donor> {}
+
+export async function getId(req: IdGenerationRequest) {
+	console.log('getId function called: ' + req.submitterDonorId);
+	const token = await getToken();
+	const headers = {
+		headers: {
+			Authorization: 'Bearer ' + token,
+		},
+	};
+
+	try {
+		const url =
+			config.getConfig().idServiceUrl() +
+			`${req.programId}/${req.submitterDonorId}/${req.submitterSpecimenId}/${req.submitterSampleId}/${req.submitterPrimaryDiagnosisId}/${req.submitterFollowUpId}/${req.submitterTreatmentId}/${req.testInterval}/${req.family_relative_id}/${req.comorbidityTypeCode}/${req.entityType}`;
+		const response = await axios.get(url, headers);
+		console.log('getId response: ' + response.data.entityId + ' - ' + response.data.entityType);
+		return parseInt(response.data.entityId);
+	} catch (e) {
+		throw new Errors.IdGenerationError('Error sending request to ID service. Caused by: ' + e);
+	}
+}
+
+export async function setEntityIdsForDonors(donors: Donor[]) {
+	const donorsWithIds = donors.map(async (donor) => {
+		return await setEntityIds(donor);
+	});
+	return await Promise.all(donorsWithIds);
+}
+
+export async function setEntityIds(donor: PartialDonor) {
+	const submitterDonorId = donor.submitterId as string;
```

Comment:
One more thing I want to take a look at, I think we can update PartialDonor like this:

```
export interface PartialDonor extends Partial<Donor> {
	submitterId: string;
	programId: string;
}
```

and then you can remove repeated use of `as string` in this function and elsewhere. These two keys are typically required keys in `Donor` so I think it fits your usage.

We try to avoid using `as` because it can create situations where you're misleading Typescript. For example, `donor.submitterId` could actually be `undefined` and TS wouldn't know.

Tagging @justincorrigible (may want to ask Jon?) because `as` usage is something we've discussed frequently before so thought I'd get a second opinion